pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java

```diff
@@ -151,6 +151,59 @@ public StreamingOutput handleTimeSeriesQueryRange(@QueryParam("language") String
     return executeTimeSeriesQueryCatching(httpHeaders, language, query, start, end, step);
   }
 
+  @POST
+  @Path("validateMultiStageQuery")
+  public MultiStageQueryValidationResponse validateMultiStageQuery(String requestJsonStr,
+      @Context HttpHeaders httpHeaders) {
+    JsonNode requestJson;
+    try {
+      requestJson = JsonUtils.stringToJsonNode(requestJsonStr);
+    } catch (Exception e) {
+      LOGGER.warn("Caught exception while parsing request {}", e.getMessage());
+      return new MultiStageQueryValidationResponse(false, "Failed to parse request JSON: " + e.getMessage(), null);
+    }
+    if (!requestJson.has("sql")) {
+      return new MultiStageQueryValidationResponse(false, "JSON Payload is missing the query string field 'sql'", null);
+    }
+    String sqlQuery = requestJson.get("sql").asText();
+    Map<String, String> queryOptionsMap = RequestUtils.parseQuery(sqlQuery).getOptions();
+    String database = DatabaseUtils.extractDatabaseFromQueryRequest(queryOptionsMap, httpHeaders);
+    try (QueryEnvironment.CompiledQuery compiledQuery = new QueryEnvironment(database,
+        _pinotHelixResourceManager.getTableCache(), null).compile(sqlQuery)) {
+      return new MultiStageQueryValidationResponse(true, null, null);
+    } catch (QueryException e) {
+      LOGGER.info("Caught exception while compiling multi-stage query: {}", e.getMessage());
+      return new MultiStageQueryValidationResponse(false, e.getMessage(), e.getErrorCode());
+    }
+  }
+
+  private static class MultiStageQueryValidationResponse {
```

Comment:
A public method should not have a private class as a return type.

Reply:
I'm only using the class for JSON serde, but fair point.